File: android/bench/Bench.java
// The same six algorithms as algorithms.ring, for Android's own runtime.
//
//     javac -> d8 -> adb push -> dalvikvm -cp bench.dex Bench
//
// WHAT THIS COMPARISON IS, AND IS NOT.
//
// It is: the same work, the same phone, the same minute, on two runtimes
// that ship on that phone. Every CHECK value must match algorithms.ring
// exactly -- that is what makes the timings comparable at all, and it is
// asserted rather than hoped for.
//
// It is NOT a claim that one language is better than the other. ART is an
// optimising compiler with a JIT and ahead-of-time profiles; Ring's VM is a
// straightforward bytecode interpreter with no JIT at all. A compiler beating
// an interpreter is the expected result, not a discovery, and the interesting
// question is only HOW FAR apart they are and WHERE the gap widens.
//
// The no-JIT property is also exactly why the Ring VM runs on Android at all
// (FINDINGS F-36): it never asks the kernel for executable memory, so the
// seccomp filter that kills other portable runtimes never fires on it. The
// thing that costs Ring speed here is the thing that bought it the platform.

public class Bench {

    static final long MOD = 1000000007L;

    public static void main(String[] args) {
        int reps = 3;
        System.out.println("Java / ART algorithm suite");
        System.out.println("==============================================");
        System.out.println("");

        long best, t;
        long v = 0;

        best = -1;
        for (int r = 0; r < reps; r++) { t = now(); v = sieve(300000); best = keep(best, now() - t); }
        report("sieve", best, v);

        best = -1;
        for (int r = 0; r < reps; r++) { t = now(); v = matmul(80); best = keep(best, now() - t); }
        report("matmul", best, v);

        best = -1;
        for (int r = 0; r < reps; r++) { t = now(); v = fib(25); best = keep(best, now() - t); }
        report("fib", best, v);

        // Built exactly as the Ring file builds it: 1-based i, (i*7919) % 100003.
        int n = 8000;
        long[] data = new long[n];
        for (int i = 1; i <= n; i++) data[i - 1] = (i * 7919L) % 100003L;
        long[] sorted = null;
        best = -1;
        for (int r = 0; r < reps; r++) { t = now(); sorted = mergeSort(data); best = keep(best, now() - t); }
        for (int i = 1; i < sorted.length; i++)
            if (sorted[i] < sorted[i - 1]) throw new RuntimeException("unsorted at " + i);
        report("mergesort", best, sum32(sorted));

        best = -1;
        for (int r = 0; r < reps; r++) { t = now(); v = binSearchAll(sorted, 6000); best = keep(best, now() - t); }
        report("binsearch", best, v);

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < 4000; i++) sb.append("the quick brown fox ");
        String big = sb.toString();
        best = -1;
        for (int r = 0; r < reps; r++) { t = now(); v = byteScan(big); best = keep(best, now() - t); }
        report("bytescan", best, v);

        System.out.println("");
        System.out.println("SUITE OK");
    }

    static long now() { return System.nanoTime(); }
    static long keep(long best, long ns) { return (best < 0 || ns < best) ? ns : best; }

    static void report(String name, long ns, long check) {
        System.out.println("CHECK " + name + " " + check);
        // Two decimals, formatted by hand so no locale can put a comma here
        // and break the campaign's parser on someone else's phone.
        long hundredths = (ns + 5000) / 10000;
        System.out.println("TIME  " + name + " " + (hundredths / 100) + "." +
                           (hundredths % 100 < 10 ? "0" : "") + (hundredths % 100));
    }

    static long sieve(int n) {
        byte[] flag = new byte[n + 1];
        long count = 0;
        for (int i = 2; i <= n; i++) {
            if (flag[i] == 0) {
                count++;
                for (long j = (long) i * i; j <= n; j += i) flag[(int) j] = 1;
            }
        }
        return count;
    }

    static long matmul(int n) {
        long[][] a = new long[n + 1][n + 1];
        long[][] b = new long[n + 1][n + 1];
        for (int i = 1; i <= n; i++)
            for (int j = 1; j <= n; j++) { a[i][j] = (i + j) % 7; b[i][j] = ((long) i * j) % 5; }
        long acc = 0;
        for (int i = 1; i <= n; i++) {
            long[] row = a[i];
            for (int j = 1; j <= n; j++) {
                long s = 0;
                for (int k = 1; k <= n; k++) s += row[k] * b[k][j];
                acc = (acc + s * (i + j)) % MOD;
            }
        }
        return acc;
    }

    static long fib(int n) { return n < 2 ? n : fib(n - 1) + fib(n - 2); }

    static long[] mergeSort(long[] l) {
        if (l.length <= 1) return l;
        int mid = l.length / 2;
        long[] left = new long[mid], right = new long[l.length - mid];
        System.arraycopy(l, 0, left, 0, mid);
        System.arraycopy(l, mid, right, 0, l.length - mid);
        return merge(mergeSort(left), mergeSort(right));
    }

    static long[] merge(long[] a, long[] b) {
        long[] out = new long[a.length + b.length];
        int i = 0, j = 0, k = 0;
        while (i < a.length && j < b.length) out[k++] = (a[i] <= b[j]) ? a[i++] : b[j++];
        while (i < a.length) out[k++] = a[i++];
        while (j < b.length) out[k++] = b[j++];
        return out;
    }

    static long binSearchAll(long[] sorted, int queries) {
        long hits = 0;
        int n = sorted.length;
        for (int q = 1; q <= queries; q++) {
            long target = (q * 7919L) % (n * 2L);
            int lo = 1, hi = n;
            while (lo <= hi) {
                int mid = (lo + hi) / 2;
                if (sorted[mid - 1] == target) { hits++; break; }
                else if (sorted[mid - 1] < target) lo = mid + 1;
                else hi = mid - 1;
            }
        }
        return hits;
    }

    static long byteScan(String s) {
        long h = 0;
        for (int i = 0; i < s.length(); i++) h = (h * 131 + s.charAt(i)) % MOD;
        return h;
    }

    static long sum32(long[] l) {
        long h = 0;
        for (int i = 0; i < l.length; i++) h = (h * 31 + l[i]) % MOD;
        return h;
    }
}

File: android/bench/bench.lua
-- The same six algorithms as algorithms.ring, for Lua 5.4.
--
--     lua bench.lua
--
-- WHY LUA IS THE FAIR COMPARISON. Android's own ART is a JIT: it compiles
-- hot loops to native arm64, so putting it beside Ring measures "compiler
-- versus interpreter" and little else. Lua 5.4 is the same KIND of thing as
-- Ring -- a small, dynamically typed, register-based bytecode interpreter
-- with no JIT, embedded rather than hosted, in the same size class. The gap
-- to Lua is therefore a gap in interpreter engineering, which is a number
-- worth knowing; the gap to ART is a gap in compilation strategy, which is
-- a different question with a different answer.
--
-- Lua is 1-indexed, like Ring, so the loops below are near-transcriptions
-- and the arithmetic is the same arithmetic. Every CHECK value must equal
-- the Ring file's exactly.

local MOD = 1000000007

local function ms(t0) return (os.clock() - t0) * 1000 end

local function report(name, msv, check)
    print(string.format("CHECK %s %d", name, check))
    -- Formatted by hand rather than with %f so no locale can decide to put
    -- a comma where the campaign's parser expects a dot.
    local h = math.floor(msv * 100 + 0.5)
    print(string.format("TIME  %s %d.%02d", name, h // 100, h % 100))
end

local function sieve(n)
    local flag = {}
    for i = 1, n do flag[i] = 0 end
    local count = 0
    for i = 2, n do
        if flag[i] == 0 then
            count = count + 1
            local j = i * i
            while j <= n do flag[j] = 1; j = j + i end
        end
    end
    return count
end

local function matmul(n)
    local a, b = {}, {}
    for i = 1, n do
        a[i] = {}; b[i] = {}
        for j = 1, n do a[i][j] = (i + j) % 7; b[i][j] = (i * j) % 5 end
    end
    local acc = 0
    for i = 1, n do
        local row = a[i]
        for j = 1, n do
            local s = 0
            for k = 1, n do s = s + row[k] * b[k][j] end
            acc = (acc + s * (i + j)) % MOD
        end
    end
    return acc
end

local function fib(n)
    if n < 2 then return n end
    return fib(n - 1) + fib(n - 2)
end

local function merge(a, b)
    local out, i, j = {}, 1, 1
    while i <= #a and j <= #b do
        if a[i] <= b[j] then out[#out + 1] = a[i]; i = i + 1
        else out[#out + 1] = b[j]; j = j + 1 end
    end
    while i <= #a do out[#out + 1] = a[i]; i = i + 1 end
    while j <= #b do out[#out + 1] = b[j]; j = j + 1 end
    return out
end

local function mergesort(l)
    if #l <= 1 then return l end
    local mid = #l // 2
    local left, right = {}, {}
    for i = 1, mid do left[#left + 1] = l[i] end
    for i = mid + 1, #l do right[#right + 1] = l[i] end
    return merge(mergesort(left), mergesort(right))
end

local function binsearchall(sorted, queries)
    local hits, n = 0, #sorted
    for q = 1, queries do
        local target = (q * 7919) % (n * 2)
        local lo, hi = 1, n
        while lo <= hi do
            local mid = (lo + hi) // 2
            if sorted[mid] == target then hits = hits + 1; break
            elseif sorted[mid] < target then lo = mid + 1
            else hi = mid - 1 end
        end
    end
    return hits
end

local function bytescan(s)
    local h = 0
    for i = 1, #s do h = (h * 131 + string.byte(s, i)) % MOD end
    return h
end

local function sum32(l)
    local h = 0
    for i = 1, #l do h = (h * 31 + l[i]) % MOD end
    return h
end

print("Lua 5.4 algorithm suite")
print("==============================================")
print("")

local reps = 3
local best, v, t0

best = -1
for _ = 1, reps do t0 = os.clock(); v = sieve(300000); local m = ms(t0); if best < 0 or m < best then best = m end end
report("sieve", best, v)

best = -1
for _ = 1, reps do t0 = os.clock(); v = matmul(80); local m = ms(t0); if best < 0 or m < best then best = m end end
report("matmul", best, v)

best = -1
for _ = 1, reps do t0 = os.clock(); v = fib(25); local m = ms(t0); if best < 0 or m < best then best = m end end
report("fib", best, v)

local n = 8000
local data = {}
for i = 1, n do data[i] = (i * 7919) % 100003 end
local sorted
best = -1
for _ = 1, reps do t0 = os.clock(); sorted = mergesort(data); local m = ms(t0); if best < 0 or m < best then best = m end end
for i = 2, #sorted do
    if sorted[i] < sorted[i - 1] then error("unsorted at " .. i) end
end
report("mergesort", best, sum32(sorted))

best = -1
for _ = 1, reps do t0 = os.clock(); v = binsearchall(sorted, 6000); local m = ms(t0); if best < 0 or m < best then best = m end end
report("binsearch", best, v)

local parts = {}
for _ = 1, 4000 do parts[#parts + 1] = "the quick brown fox " end
local big = table.concat(parts)
best = -1
for _ = 1, reps do t0 = os.clock(); v = bytescan(big); local m = ms(t0); if best < 0 or m < best then best = m end end
report("bytescan", best, v)

print("")
print("SUITE OK")
